[SofaHelper SofaGraphComponent] issofa_plugin: Plugin management

fjourdes opened the pull request:
I don't believe there was a pull request already attached to issofa_plugins branch ? My apologies if that 
was the case. 

# Purpose
Some basic plugin functionalities were silently changed for no good reasons other the past years, namely
- SOFA_LIBSUFFIX  which should be set as "_d" during a debug build is broken under msvc since c89da68 which as the commit name says is indeed monstruous. It was "fixed" later by adding directly inside the PluginManager code a condition to add the "_d" suffix by checking the compiler. 
- The default behavior of the global `PluginRepository`, was changed by c5a19fe for absolutely no good reasons.
It breaks a very basic functionaliy ( you can always load a library that is in your working directory ), also this change had some **interesting** side effects when basically you write a unit test which relies on
a scene file that uses a `RequiredPlugin`... that shortcoming was somewhat hidden by having all tests relying on `SofaGTest`, whose sole reason of being is to actually amend c5a19fe by populating the PluginRepository with the current working dir.

Some additional functionalities are added to `RequiredPlugin` by this PR, namely the ability to specify other suffixes ( or no suffix at all ) to prepend when attemtping to load a plugin with a name hint. 

# CHANGELOG 
FIX
[SofaHelper]
- Restore the `SOFA_LIBSUFFIX` behavior with msvc
- `PluginRepository` is initialized to contain the current working directory

CHANGE
[SofaHelper] 
- `PluginManager` : keep only a single version of the `loadPlugin` method. Also the method returns true
even if the plugin is already loaded. Returning false is kept only for real failures, whereby the library could 
not be loaded. If it is important to check wether or not a plugin is already loaded, the result of the `pluginIsLoaded` method can be inspected beforehand. 

[SofaGraphComponent] 
- `RequiredPlugin` component gives the ability to override the default suffix which is "_d" in debug mode.  
It also gives some more advanced options to extend the plugin search.

# Remarks

I do not know which version of the GUI is still in use. However I believe that the widget that can add/remove plugins to the `PluginManager` should be removed since there is no valid reason to 
keep it.

I am really not sure about the current state of SofaGTestMain. It is hidden somewhere in the folder hierarchy (inside tools directory), and its existence seems to imply that you cannot write a unit test inside sofa without relying on it to "initialise Sofa" which is absolutely not the case. See PR #471
Instead having a library which depends on gtest and SofaCore which 
- factorize some initialisation methods when they are actually required in the test `SetUp()` method
- provide some useful abstractions when you want to test some important concepts of a component, like what SofaTest wanted to do, but without the dependency bloat. A potential non exhaustive list of abstactions such a library could provide 
  - Mapping 
     - linearisation, 
     - applyJ applyJT equivalence 
  - ForceField
    - linearisation
    - addDForce addKToMatrix equivalence 
    - addKToMatrix must give a SPD matrix (well actually in Sofa convention I think internal forcefields actually give a SND matrix but the idea remains...

Nothing of this testing API would require the initialisation of a simulation / node.

   
 ______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

JeremieA committed:
CHANGE PluginManager API to allow for custom library suffixes and allow for more flexible error handling. ADD suffixMap Data in RequiredPlugin and change name Data into a vector to support loading of plugins with non standard version/debug suffixes and several alternate names or specifying a group of plugins to load

git-svn-id: https://code.insimo.fr/svn/IS/ISSofa/branches/ISSofa@10325 f8a8aa39-ebdb-4476-bae3-08f188f1b544

Conflicts:
	framework/sofa/helper/system/PluginManager.cpp
	modules/sofa/component/misc/RequiredPlugin.cpp
	modules/sofa/component/misc/RequiredPlugin.h

fjourdes committed:
FIX COMPILATION: SofaGraphComponent

fjourdes committed:
FIX: RequiredPlugin logging

Commit:
FIX: compilation error

Commit:
[SofaGraphComponent] Fix merge of issofa improvement. Commit 69d911016adc80616b453587a05f3ef07cb9a594 CHANGE PluginManager API to allow for custom library suffixes and allow for more flexible error handling. Keep the possibility of loading plugin by name if PluginName is empty introduced in master.

fjourdes committed:
[SofaHelper] FIX: msvc SOFA_LIBSUFFIX preprocessor macro remains empty
in debug mode. Since Msvc generates multi configuration makefiles
(Debug, RelWithDebug, Release), the CMAKE_BUILD_TYPE is therefore never
equal to Debug. The expected behavior was broken by c89da68

fjourdes committed:
[SofaHelper] CHANGE: PluginManager loadPlugin method signature.
Changed also the behavior whereby the loadPlugin method returns true
even if the plugin is already loaded, which is therefore no longer
considered as a failure.
In codes where it matters to check if a plugin is already loaded or
not, the result of the pluginIsLoaded method can be checked before any
further call to loadPlugin.

fjourdes committed:
[SofaHelper] CHANGE: PluginRepository should not be empty by default,
but should contain the current working directory instead.
This was the (natural) behavior before c5a19fe6, however this change had
the unwanted side effect to break for example, any unit test executable
that attempted to load a plugin in its working directory, since it made the
PluginRepository empty by default. There is absolutely no reason to make
it the default behavoir, since any operating system, by default, adds the
current working directory to the PATH environment of the exectuable
when launching it.

fjourdes committed:
[SofaGraphComponent] FIX: RequiredPlugin compilation with respect to
changes in PluginManager::loadPlugin method signature.

fjourdes committed:
UPDATE: runSofa no longer needs to populate the PluginRepository with its
working directory.

fjourdes committed:
Merge remote-tracking branch 'sofa_github/master' into issofa_plugins

fjourdes committed:
[runSofa] MERGE FIX: retrieve plugin directory using getFirstPath method
from PluginRepository.

fjourdes committed:
[SofaGuiQt] UPDATE: SofaPluginManager to call loadPluginByPath method.

fjourdes committed:
[SofaGraphComponent] MERGE FIX: pluginName -> d_pluginName

fredroy commented:
I totally agree on the fact that the PluginManager widget in the Qt GUI is irrelevant now (useless to load a plugin manually AFTER loading the scene using element from it).

fredroy commented on SofaKernel/framework/sofa/helper/system/PluginManager.cpp:
```diff
@@ -289,12 +296,14 @@ std::string PluginManager::findPlugin(const std::string& pluginName, bool ignore
-                if (downcaseFilename == downcaseLibName)
+                if (downcaseFilename == downcaseLibName) {
                     return dir + "/" + filename;
+                    std::cout<<"dir : "<<dir + "/" + filename<<std::endl;
```
message API (msg_error/warning/etc) should be used here

fjourdes replied:
This likely result from the merging. The std::cout statement will in fact never been called, since it is after a return statement. The line should be removed entirely.

fredroy commented on SofaKernel/framework/sofa/helper/system/PluginManager.h:
```diff
@@ -136,15 +136,41 @@ class SOFA_HELPER_API PluginManager
+	bool loadPluginByName(const std::string& pluginName, const std::string& suffix = getDefaultSuffix(), bool ignoreCase = true, std::ostream* errlog= nullptr);
+    
+	/// Unloads a plugin from process memory.
+	bool unloadPlugin(const std::string& path, std::ostream* errlog= nullptr);
```
Nice comments !

fredroy commented on modules/SofaGraphComponent/RequiredPlugin.cpp:
```diff
@@ -42,48 +42,93 @@ int RequiredPluginClass = core::RegisterObject("Load the required plugins")
+        {
+            sout << errmsg.str();
+            sout << "Optional/alternate plugin"<<(failed.size()>1?"s":"")<<" failed to load: " << failed << sendl;
+        }
```
same as before, about the message API

fjourdes replied:
This one comes from our local version. We do not use the message API at InSimo

fjourdes committed:
[SofaHelper] CLEAN: PluginManager
- documentation of the loadPlugin* methods
- removed never executed code after return statement
- replaced tabs by spaces

fredroy commented:
[ci-build]

fredroy committed:
Update RequiredPlugin.cpp
